### PR TITLE
feat: add auth_mode for explicit auth configuration

### LINKED ISSUE #1877: [FEAT]: Switching Between PAT and GitHub App Authentication Without Modifying Terraform Code
### Describe the need

Hello,

I'm trying to differentiate between authentication methods: using PAT (Personal Access Token) in my local environment and the GitHub App in the CI environment. However, after adding the app_auth block for the CI setup, I receive an error in the local environment indicating that id, installation_id, and pem_file are not set. Is there a way to toggle between authentication methods without changing the Terraform code?

Thank you for your assistance.

(Note: This message was translated with the assistance of a machine translation tool.)

### SDK Version

_No response_

### API Version

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #3117: [MAINT]: Add provider configuration to set `auth_mode: 'anonymous'`
### Describe the need

We want to ensure that users can explicitly disable auth. Then `terraform` doesn't need to do extra work to check all other auth types.

### SDK Version

_No response_

### API Version

_No response_

### Relevant log output

```shell

```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by laughedelic**
This PR 
- Resolves https://github.com/integrations/terraform-provider-github/issues/1877
  - Supersedes https://github.com/integrations/terraform-provider-github/pull/2174
- Resolves https://github.com/integrations/terraform-provider-github/issues/3117
- Partially addresses https://github.com/integrations/terraform-provider-github/issues/3116 
  - see the note at the bottom on why it doesn't resolve that issue
- Also, partially addresses https://github.com/integrations/terraform-provider-github/issues/598 (improved error when `auth_mode` is set explicitly)

----

### Before the change?

- `app_auth {}` block is the only way to configure GitHub App authentication; switching between token and app auth requires modifying provider configuration code
- When using environment variables for app auth, an empty `app_auth {}` block is required due to a Terraform SDK limitation (#1877)
- Authentication mode is implicitly auto-detected with no way to opt out: if `GITHUB_TOKEN` is in the environment or `gh` CLI has a valid session, anonymous mode is impossible (#3116)

### After the change?

Based on the proposal in #3116:

- New `auth_mode` argument enables explicit control over authentication:
  - `auth_mode = "anonymous"` ignores any credentials
  - `auth_mode = "token"` errors if no token is provided instead of silently falling back
  - `auth_mode = "app"` errors if app credentials are missing or incomplete
  - When `auth_mode` is not set, the existing auto-detection behavior is preserved (backward compatible)
- New top-level `app_id`, `app_installation_id`, `app_private_key` arguments allow app auth without the `app_auth` block, making it possible to switch between token and app auth purely via environment variables (addressing #1877)
- `app_auth` block is deprecated with a mention of the new top-level arguments
- `gh auth token` CLI fallback is deprecated (as proposed in #3116)

Note, I took the liberty to name the new top-level argument `app_private_key` instead of `app_pem_file` because I believe it is more clear that the expected value is a private key content, not a file reference. I remember there was this [old comment](https://github.com/integrations/terraform-provider-github/issues/949#issuecomment-1743755325):

> I feel like the naming of the `pem_file` argument in `app_auth` is a little confusing because it expects the content of the file, not its path. Maybe calling it `private_key` would have been clearer

---

### Pull request checklist

- [x] Schema migrations have been created if needed ([example](https://github.com/F-Secure-web/terraform-provider-github/blob/main/github/migrate_github_actions_organization_secret.go))
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?

- [x] No

----

I didn't implement `environment_variable_prefix` and `auth_type` from the #3116's proposal to keep this focused, and implementing those seemed like touching many more places. But I can add those separately.

- Follow-up: https://github.com/laughedelic/terraform-provider-github/pull/1

### CODE REVIEW

File: github/provider.go [hunk 1]
```diff
@@ -419,11 +456,42 @@ func providerConfigure(p *schema.Provider) schema.ConfigureContextFunc {
...
+				log.Printf("[INFO] No token found, trying GitHub CLI to get token from hostname %s", baseURL.Host)
+				ghToken := tokenFromGHCLI(baseURL)
+				if ghToken != "" {
+					token = ghToken
+					diags = append(diags, diag.Diagnostic{
+						Severity: diag.Warning,
+						Summary:  "GitHub CLI token fallback is deprecated",
+						Detail: "Automatic token detection from `gh auth token` is deprecated and will be removed in a future major release. " +
+							"Please set the `token` provider argument or `GITHUB_TOKEN` environment variable explicitly. " +
+							"You can use `export GITHUB_TOKEN=$(gh auth token)` as a replacement.",
```
Comment (laughedelic) [inline github/provider.go:510]: I've added a deprecation for gh CLI as proposed in #3116, but another option could be adding an explicit `auth_mode = cli` instead

**Comment by laughedelic** [comment]
Thinking some more about the `auth_mode` and the empty default with some auto-resolution for compatibility: what if it was a priority list instead of a single value? This would allow users to set their own fallback logic by saying `auth_mode = ["app", "cli", "token"]` and depending on the environment, it would use different options in that order. WDYT?

UPD: Maybe I'm just overcomplicating it. Feel free to ignore this.

File: github/provider.go (hunk 1 above)
Comment (laughedelic) [reply github/provider.go:510]: Never mind. It's probably not worth it.

File: github/provider.go [hunk 2]
```diff
@@ -378,34 +412,37 @@ func providerConfigure(p *schema.Provider) schema.ConfigureContextFunc {
...
 		}
 
-		if appAuth, ok := d.Get("app_auth").([]any); ok && len(appAuth) > 0 && appAuth[0] != nil {
-			appAuthAttr := appAuth[0].(map[string]any)
+		var token string
 
-			var appID, appInstallationID, appPemFile string
+		switch authMode {
+		case "anonymous":
+			log.Printf("[INFO] Auth mode: anonymous")
```
Comment (deiga) [inline github/provider.go]: Could you test if we can use `tflog` inside this func?

File: github/provider.go [hunk 3]
```diff
@@ -378,34 +412,37 @@ func providerConfigure(p *schema.Provider) schema.ConfigureContextFunc {
...
 
-			if v, ok := appAuthAttr["id"].(string); ok && v != "" {
-				appID = v
-			} else {
-				return nil, wrapErrors([]error{fmt.Errorf("app_auth.id must be set and contain a non-empty value")})
+		case "token":
+			token = d.Get("token").(string)
+			if token == "" {
+				return nil, diag.FromErr(fmt.Errorf(
```
Comment (deiga) [inline github/provider.go]:
```suggestion
				return nil, diag.Errorf(
```

File: github/provider.go [hunk 4]
```diff
@@ -123,6 +130,25 @@ func Provider() *schema.Provider {
...
+				DefaultFunc: schema.EnvDefaultFunc("GITHUB_APP_INSTALLATION_ID", nil),
+				Description: descriptions["app_installation_id"],
+			},
+			"app_private_key": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Sensitive:   true,
+				DefaultFunc: schema.EnvDefaultFunc("GITHUB_APP_PRIVATE_KEY", nil),
+				Description: descriptions["app_private_key"],
+			},
```
Comment (deiga) [inline github/provider.go:156]: Could you add `ConflictsWith` fields so that it's not possible to set both App auth configs at the same time? Or does that interfere with the `DefaultFunc` being the same?

File: github/provider_test.go [hunk 5]
```diff
@@ -232,6 +232,135 @@ data "github_ip_ranges" "test" {}
...
+			Steps: []resource.TestStep{
+				{
+					Config:      config,
+					ExpectError: regexp.MustCompile(`auth_mode is set to "app" but the following app credentials are missing`),
+				},
+			},
+		})
+	})
+
+	t.Run("can be configured with top-level app fields without auth_mode", func(t *testing.T) {
```
Comment (deiga) [inline github/provider_test.go]: Hmm, I would prefer if this wouldn't be possible. Can we force setting of `auth_mode` for top-level app fields? Or are we limited by backwards compatibility?

**Review by deiga — CHANGES_REQUESTED** [review]
I think this is looking great!

I've been wanting to get to this for a while now 😬 

Left a few minor comments

File: github/provider.go (hunk 2 above)
Comment (laughedelic) [reply github/provider.go]: Seems to be fine. Done in e07604b.

File: github/provider.go (hunk 3 above)
Comment (laughedelic) [reply github/provider.go]: Done in c33ccb5

File: github/provider.go (hunk 4 above)
Comment (laughedelic) [reply github/provider.go:156]: I don't think it interferes with the default value, but it probably doesn't raise the conflict when everything is set through env vars. In any case, I added mutually exclusive ConflictsWith in e3264c4. And further validation is done in `providerConfigure`.

File: github/provider_test.go (hunk 5 above)
Comment (laughedelic) [reply github/provider_test.go]: This is a bit more involved than just requiring `auth_mode = app` if the top-lvel app fields are set. I added an extra validation in the "default" case, to make sure that if the top-level fields are set and there is no `app_auth` block, we get an error about setting `auth_mode`.
There are cases when user can set both the token and the app values, but then choose which set of credentials to use via `auth_mode`. Any or all of them can be set via env vars. So I think this shouldn't be too strict. 
See af6fb59. I added some extra tests for the edge cases. Let me know if we're on the same page.

**Review by deiga — APPROVED** [review]
Whoo! This seems so much simpler than what I had thought it would be 😬 

I hope we're not missing something

**Comment by laughedelic** [comment]
@stevehipwell could you give some feedback here?

**Comment by stevehipwell** [comment]
Thanks for the PR @laughedelic, this whole area changed significantly in #3448 and we're waiting until v7 to cleanup the unexpected behaviours. That all said I'd be interested in getting in the `auth_mode` configuration option in (maybe with `legacy` as the default option).

**Comment by laughedelic** [comment]
Closing. `auth_mode` landed in #3578. For the record, this PR proposed the same design back in March and was approved here (two years after the first attempt in #2174). Not a great contributing experience.

Glad the feature is in either way, and #1877 seems to be actually fixed.

**Comment by deiga** [comment]
@laughedelic I'm sorry about this. It's hard to keep track of all the contributed work, there is so much. 
And we're just a few people who maintain this for a hobby, there isn't a lot of capacity to make sure that the contributing experience is great.

**Comment by laughedelic** [comment]
@deiga I understand. Thanks for the response and your previous reviews!
